features/bsm/service/impl/src/main/java/org/opennms/netmgt/bsm/service/internal/DefaultBusinessServiceStateMachine.java

```diff
@@ -213,20 +226,22 @@ private void reduceUpdateAndPropagateVertex(BusinessServiceGraph graph, GraphVer
      * @param edgesWithStatus
      * @return
      */
-    public static List<Status> weighStatuses(Map<GraphEdge, Status> edgesWithStatus) {
+    public static List<StatusWithIndex> weighStatuses(Map<GraphEdge, Status> edgesWithStatus) {
         // Find the greatest common divisor of all the weights
         int gcd = edgesWithStatus.keySet().stream()
                 .map(e -> e.getWeight())
                 .reduce((a,b) -> BigInteger.valueOf(a).gcd(BigInteger.valueOf(b)).intValue())
                 .orElse(1);
 
         // Multiply the statuses based on their relative weight
-        List<Status> statuses = Lists.newArrayList();
+        List<StatusWithIndex> statuses = Lists.newArrayList();
+        int k = 0;
         for (Entry<GraphEdge, Status> entry : edgesWithStatus.entrySet()) {
             int relativeWeight = Math.floorDiv(entry.getKey().getWeight(), gcd);
             for (int i = 0; i < relativeWeight; i++) {
-                statuses.add(entry.getValue());
+                statuses.add(new StatusWithIndex(entry.getValue(), k));
```

Comment:
This looks fishy as it generates the index by the position of an element in a set.

Is there any mechanism ensuring the set does not change while this graph is active?